### PR TITLE
Pass Verbosity to all components

### LINKED ISSUE #288: [kwokctl] Passing the `-v` level to all components
### What would you like to be added?

- kwok-controller

Need to convert to the component's own `-v` level
- etcd
- kube-apiserver
- kube-controller-manager
- kube-schduler
- prometheus

### Why is this needed?

NONE

### PR DESCRIPTION
**Pull request by Zhuzhenghao**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

Fixes #288

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by wzshiming** [comment]
Can we have a test cover this change?

**Comment by Zhuzhenghao** [comment]
Work on e2e test cases.

File: pkg/log/wrapper.go [hunk 1]
```diff
@@ -76,11 +85,35 @@ func (l *Logger) Error(msg string, err error, args ...any) {
...
+	return wrapSlog(l.log.With(args...), l.level)
 }
 
 // WithGroup returns a new Logger that starts a group. The keys of all
 // attributes added to the Logger will be qualified by the given name.
 func (l *Logger) WithGroup(name string) *Logger {
-	return wrapSlog(l.log.WithGroup(name))
+	return wrapSlog(l.log.WithGroup(name), l.level)
+}
+
```
Comment (wzshiming) [inline pkg/log/wrapper.go:96]:
```suggestion

func (l *Logger) Level() int {
	return l.level
}

```

File: pkg/log/wrapper.go [hunk 2]
```diff
@@ -76,11 +85,35 @@ func (l *Logger) Error(msg string, err error, args ...any) {
...
+// ToKlogLevel maps the current logging level to a Klog level integer
+func (l *Logger) ToKlogLevel() int {
```
Comment (wzshiming) [inline pkg/log/wrapper.go:98]:
```suggestion
func ToKlogLevel(level int) int {
```

File: pkg/log/wrapper.go [hunk 3]
```diff
@@ -76,11 +85,35 @@ func (l *Logger) Error(msg string, err error, args ...any) {
...
+	if int(l.level) > 0 {
+		return 0
+	}
+	return -int(l.level)
+}
+
+// ToLogSeverityLevel maps the current logging level to a severity level string
+func (l *Logger) ToLogSeverityLevel() string {
```
Comment (wzshiming) [inline pkg/log/wrapper.go:106]:
```suggestion
func ToLogSeverityLevel(level int) string {
```

File: pkg/log/wrapper.go (hunk 2 above)
Comment (wzshiming) [reply pkg/log/wrapper.go:98]: move to new file

File: pkg/log/wrapper.go (hunk 3 above)
Comment (wzshiming) [reply pkg/log/wrapper.go:106]: move to new file

File: pkg/log/wrapper.go [hunk 4]
```diff
@@ -32,13 +32,22 @@ const (
...
 
-func wrapSlog(log *slog.Logger) *Logger {
-	return &Logger{log}
+// The following is Level security definitions.
+const (
+	InfoLevelSecurity  = "info"
+	DebugLevelSecurity = "debug"
+	WarnLevelSecurity  = "warn"
+	ErrorLevelSecurity = "error"
+)
```
Comment (wzshiming) [inline pkg/log/wrapper.go:41]: move to new file

File: pkg/apis/v1alpha1/kwokctl_configuration_types.go [hunk 5]
```diff
@@ -269,6 +276,9 @@ type KwokctlConfigurationOptions struct {
 
 	// CacheDir is the directory of the cache.
 	CacheDir string `json:"cacheDir,omitempty"`
+
+	// Log is logging level, the value of the -v flag
+	Log
```
Comment (wzshiming) [inline pkg/apis/v1alpha1/kwokctl_configuration_types.go:281]:
```suggestion
```

File: pkg/apis/v1alpha1/kwokctl_configuration_types.go [hunk 6]
```diff
@@ -43,6 +43,13 @@ type KwokctlConfiguration struct {
 	Components []Component `json:"components,omitempty"`
 }
 
+// Log returns
+type Log struct {
+	// KlogLevel is logging level, the value of the -v flag
+	KlogLevel     int
+	SeverityLevel string
+}
+
```
Comment (wzshiming) [inline pkg/apis/v1alpha1/kwokctl_configuration_types.go:52]:
```suggestion
```

File: pkg/kwokctl/cmd/create/cluster/cluster.go [hunk 7]
```diff
@@ -43,8 +43,13 @@ type flagpole struct {
 
 // NewCommand returns a new cobra.Command for cluster creation
 func NewCommand(ctx context.Context) *cobra.Command {
+	logger := log.FromContext(ctx)
 	flags := &flagpole{}
 	flags.KwokctlConfiguration = config.GetKwokctlConfiguration(ctx)
+	flags.Options.Log = internalversion.Log{
+		KlogLevel:     logger.ToKlogLevel(),
+		SeverityLevel: logger.ToLogSeverityLevel(),
+	}
```
Comment (wzshiming) [inline pkg/kwokctl/cmd/create/cluster/cluster.go:52]:
```suggestion
	flags.Options.Log = internalversion.Log{
		KlogLevel:     logger.ToKlogLevel(),
		SeverityLevel: logger.ToLogSeverityLevel(),
	}
```

Move to the installation phase

**Review by wzshiming — CHANGES_REQUESTED** [review]

File: pkg/kwokctl/components/etcd.go [hunk 8]
```diff
@@ -112,6 +114,10 @@ func BuildEtcdComponent(conf BuildEtcdComponentConfig) (component internalversio
 		)
 	}
 
+	if conf.Version.GT(version.NewVersion(3, 3, 17)) && conf.LogLevel != log.InfoLevelSecurity {
```
Comment (wzshiming) [inline pkg/kwokctl/components/etcd.go:117]: This flag was added by 3.3.17? Shouldn't it be 3.4 or 3.3?

File: pkg/kwokctl/runtime/kind/kind.yaml.tpl [hunk 9]
```diff
@@ -33,7 +33,7 @@ nodes:
 
   kubeadmConfigPatches:
 
-{{ if or .EtcdExtraArgs .EtcdExtraVolumes }}
+{{ if or .EtcdExtraArgs .EtcdExtraVolumes .Verbosity }}
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/kind/kind.yaml.tpl:36]: Move to https://github.com/kubernetes-sigs/kwok/blob/730d50e32b7968c4377b285f97c67d4819e8b657/pkg/kwokctl/runtime/kind/kind.go#L47

File: test/kwokctl/kwokctl_binary_verbosity.test.sh [hunk 10]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
+# Copyright 2022 The Kubernetes Authors.
```
Comment (wzshiming) [inline test/kwokctl/kwokctl_binary_verbosity.test.sh:2]:
```suggestion
# Copyright 2023 The Kubernetes Authors.
```

File: test/kwokctl/kwokctl_docker_verbosity.test.sh [hunk 11]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
+# Copyright 2022 The Kubernetes Authors.
```
Comment (wzshiming) [inline test/kwokctl/kwokctl_docker_verbosity.test.sh:2]:
```suggestion
# Copyright 2023 The Kubernetes Authors.
```

File: pkg/kwokctl/components/etcd.go [hunk 12]
```diff
@@ -114,6 +116,10 @@ func BuildEtcdComponent(conf BuildEtcdComponentConfig) (component internalversio
 		)
 	}
 
+	if conf.Version.GTE(version.NewVersion(3, 4, 0)) && conf.LogLevel != log.InfoLevelSecurity {
+		etcdArgs = append(etcdArgs, "--log-level="+conf.LogLevel)
```
Comment (wzshiming) [inline pkg/kwokctl/components/etcd.go:120]: https://etcd.io/docs/v3.5/upgrades/upgrade_3_4/#deprecating-etcd---debug-flag-now---log-leveldebug

If etcd is less than 3.4 and the logs level is debug, then we use `--debug`

**Comment by wzshiming** [comment]
```
error validating "pkg/log/convert.go": the file is missing a boilerplate
```

**Review by wzshiming — APPROVED** [review]
/approve
/lgtm

**Pull request merged**